File: mergesort/merge_sort_step.java
import java.io.*;
import java.util.*;

public class merge_sort_step {
    public static List<String> steps= new ArrayList<>();
    public static void main(String[] args){
        // see whether got 3 arguments
        if(args.length != 3){
            System.out.println("command line: java MergeSortWithStep <filename> <start> <end>");
            return;
        }

        // get all value from command line
        String fileName = args[0];
        int start = Integer.parseInt(args[1]);//assign as integer value from a string
        int end = Integer.parseInt(args[2]);
        try {
            List<String[]> list= readCSVFileToGetValue(fileName, start, end);
            mergeSort(list,0,list.size()-1);
            writeStepsToFile("merge_sort_step_"+start+"_"+end+".txt");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static List<String[]> readCSVFileToGetValue(String fileName,int start,int end) throws IOException{
        // initialize all needed variable
        List<String[]> dataToReturn = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(fileName));
        String line;
        int i=0;
        // loop to see whether any next line is empty
        while((line=reader.readLine())!=null){
            // if index is in range
            if((i>=start) && (i<=end)){
                //store the value by split with "," example "123,Elson" become {"123","Elson"} 
                String[] values = line.split(",");
                // add into data
                dataToReturn.add(values);
            }
            if(i>end){
                //stop if index more than end
                break;
            }
            i++;
        }
        reader.close();
        //return data
        return dataToReturn;

    }

    public static void mergeSort(List<String[]> list, int leftIndex, int rightIndex){
        if(leftIndex<rightIndex){
            // get mid index value
            int midIndex=(leftIndex+rightIndex)/2;
            //recursion
            mergeSort(list,leftIndex,midIndex);
            mergeSort(list, midIndex+1, rightIndex);
            //merge function call to sort after cannot recur any more
            merge(list,leftIndex,midIndex,rightIndex);
            // pass the current list after merge
            logCurrentStep(list);
        }
    }

    public static void merge(List<String[]> list,int leftIndex, int midIndex, int rightIndex){
        //left part
        List<String[]> leftList= new ArrayList<>(list.subList(leftIndex, midIndex+1));
        //right part
        List<String[]> rightList = new ArrayList<>(list.subList(midIndex+1, rightIndex+1));
        // initialize needed variable
        int i = 0, j = 0, k = leftIndex;
        // if both side i left and j right still lower than size of
        while (i < leftList.size() && j < rightList.size()) {
            // get left and right available int value
            int val1 = Integer.parseInt(leftList.get(i)[0]);
            int val2 = Integer.parseInt(rightList.get(j)[0]);
            // if left <= right will store left to the list else will store right into list
            if (val1 <= val2) {
                list.set(k++, leftList.get(i++));
            } else {
                list.set(k++, rightList.get(j++));
            }
        }
        // if left or right either one no more to store will straight set into the list, start from left first then right
        while (i < leftList.size()) {
            list.set(k++, leftList.get(i++));
        }
        while (j < rightList.size()) {
            list.set(k++, rightList.get(j++));
        }
    }

    public static void logCurrentStep(List<String[]> list) {
        StringBuilder sb = new StringBuilder();
        for (String[] entry : list) {
            // add all the entry to a line of string with <int value>/<string value>,
            sb.append(entry[0]).append("/").append(entry[1]).append(", ");
        }
        // Remove last comma and space
        if (sb.length() >= 2) sb.setLength(sb.length() - 2);
        //store into steps
        steps.add(sb.toString());
    }

    public static void writeStepsToFile(String filename) throws IOException {
        BufferedWriter writer = new BufferedWriter(new FileWriter(filename));
        for (String step : steps) {
            writer.write("["+step+"]");
            writer.newLine();
        }
        writer.close();
        System.out.println("Steps written to: " + filename);
    }
}

File: mergesort/MergeSortStep.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

using namespace std;

struct Entry {
    int number;
    string word;
};

string* steps = nullptr;  // dynamic step log array
int stepCount = 0;
int stepCapacity = 0;

void addStep(string step) {
    if (stepCount >= stepCapacity) {
        // Resize step log if needed
        stepCapacity = stepCapacity == 0 ? 10 : stepCapacity * 2;
        string* newSteps = new string[stepCapacity];
        for (int i = 0; i < stepCount; ++i)
            newSteps[i] = steps[i];
        delete[] steps;
        steps = newSteps;
    }
    steps[stepCount++] = step;
}

Entry* readCSV(const string& filename, int start, int end, int& outSize) {
    ifstream file(filename);
    string line;
    int row = 0, count = 0;
    Entry* entries = new Entry[end - start + 1];

    while (getline(file, line)) {
        if (row >= start && row <= end) {
            stringstream ss(line);
            string numStr, word;
            getline(ss, numStr, ',');
            getline(ss, word);
            entries[count++] = { stoi(numStr), word };
        }
        if (row > end) break;
        row++;
    }
    outSize = count;
    return entries;
}

void logStep(Entry arr[], int size) {
    string line;
    for (int i = 0; i < size; ++i) {
        line += to_string(arr[i].number) + "/" + arr[i].word + ", ";
    }
    if (!line.empty()) line = line.substr(0, line.size() - 2);
    addStep(line);
}

void merge(Entry arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    Entry* L = new Entry[n1];
    Entry* R = new Entry[n2];

    for (int i = 0; i < n1; ++i) L[i] = arr[left + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    while (i < n1 && j < n2) {
        if (L[i].number <= R[j].number)
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];

    delete[] L;
    delete[] R;
}

void mergeSort(Entry arr[], int left, int right, int fullSize) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(arr, left, mid, fullSize);
        mergeSort(arr, mid + 1, right, fullSize);
        merge(arr, left, mid, right);
        logStep(arr, fullSize);
    }
}

void writeStepsToFile(const string& filename) {
    ofstream out(filename);
    for (int i = 0; i < stepCount; ++i) {
        out << steps[i] << "\n";
    }
    out.close();
}

int main(int argc, char* argv[]) {
    if (argc != 4) {
        cout << "Usage: ./merge_sort_step_array <filename> <startRow> <endRow>\n";
        return 1;
    }

    string filename = argv[1];
    int start = stoi(argv[2]);
    int end = stoi(argv[3]);
    int size;

    Entry* data = readCSV(filename, start, end, size);
    mergeSort(data, 0, size - 1, size);

    string outFile = "merge_sort_step_" + to_string(start) + "_" + to_string(end) + ".txt";
    writeStepsToFile(outFile);

    cout << "Steps written to: " << outFile << endl;

    delete[] data;
    delete[] steps;
    return 0;
}
